tests/test_core.cpp

```diff
@@ -102,3 +106,165 @@ TEST(CoreTests, AdvancedMeshRemove) {
+        {{0, 0, 0}, {1, 1, 1}, {2, 2, 2}, {3, 3, 3}, {4, 4, 4}}};
+
+    di_renderer::core::Mesh mesh(vertices, tex_coords, normals, faces);
+
+    EXPECT_EQ(mesh.face_count(), 3);
+
+    auto triangle1 = mesh.triangulated_faces[0];
+    EXPECT_EQ(triangle1[0], 0);
+    EXPECT_EQ(triangle1[1], 1);
+    EXPECT_EQ(triangle1[2], 2);
+
+    auto triangle2 = mesh.triangulated_faces[1];
+    EXPECT_EQ(triangle2[0], 0);
+    EXPECT_EQ(triangle2[1], 2);
+    EXPECT_EQ(triangle2[2], 3);
+
+    auto triangle3 = mesh.triangulated_faces[2];
+    EXPECT_EQ(triangle3[0], 0);
+    EXPECT_EQ(triangle3[1], 3);
+    EXPECT_EQ(triangle3[2], 4);
+}
+
+TEST(MeshTriangulationTest, SingleFaceWithLessThanThreeVerticesIgnored) {
+    std::vector<di_renderer::math::Vector3> vertices = {{0, 0, 0}, {1, 0, 0}};
+    std::vector<di_renderer::math::UVCoord> tex_coords = {{0, 0}, {1, 0}};
+    std::vector<di_renderer::math::Vector3> normals = {{0, 0, 1}, {0, 0, 1}};
+
+    std::vector<std::vector<di_renderer::core::FaceVerticeData>> faces = {{
+        {0, 0, 0}, {1, 1, 1} // Only 2 vertices
+    }};
+
+    di_renderer::core::Mesh mesh(vertices, tex_coords, normals, faces);
+
+    EXPECT_EQ(mesh.face_count(), 0);
+}
+
+TEST(MeshTriangulationTest, MultipleFacesAllTriangulated) {
+    std::vector<di_renderer::math::Vector3> vertices = {{0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, {2, 0, 0}};
+    std::vector<di_renderer::math::UVCoord> tex_coords = {{0, 0}, {1, 0}, {1, 1}, {0, 1}, {2, 0}};
+    std::vector<di_renderer::math::Vector3> normals = {{0, 0, 1}, {0, 0, 1}, {0, 0, 1}, {0, 0, 1}, {0, 0, 1}};
+
+    std::vector<std::vector<di_renderer::core::FaceVerticeData>> faces = {
+        {{0, 0, 0}, {1, 1, 1}, {2, 2, 2}},           // Triangle
+        {{0, 0, 0}, {1, 1, 1}, {2, 2, 2}, {3, 3, 3}} // Quad
+    };
+
+    di_renderer::core::Mesh mesh(vertices, tex_coords, normals, faces);
+
+    EXPECT_EQ(mesh.face_count(), 3); // 1 triangle + 2 from quad = 3
+}
+
+TEST(MeshNormalComputationTest, ProvidedNormalsArePreserved) {
+    std::vector<di_renderer::math::Vector3> vertices = {di_renderer::math::Vector3(0.0F, 0.0F, 0.0F),
+                                                        di_renderer::math::Vector3(1.0F, 0.0F, 0.0F),
+                                                        di_renderer::math::Vector3(0.0F, 1.0F, 0.0F)};
+
+    std::vector<di_renderer::math::Vector3> provided_normals = {di_renderer::math::Vector3(0.0F, 0.0F, 1.0F),
+                                                                di_renderer::math::Vector3(0.0F, 0.0F, 1.0F),
+                                                                di_renderer::math::Vector3(0.0F, 0.0F, 1.0F)};
+
+    std::vector<std::vector<di_renderer::core::FaceVerticeData>> faces = {{{0, -1, -1}, {1, -1, -1}, {2, -1, -1}}};
+
+    Mesh mesh(std::move(vertices), {}, std::move(provided_normals), faces);
+
+    ASSERT_EQ(mesh.normals.size(), 3U);
+    EXPECT_NEAR(mesh.normals[0].x, 0.0F, 1e-5F);
```

Comment:
Здесь и далее по рекомендации Косенко
```suggestion
    EXPECT_NEAR(mesh.normals[0].x, 0.0F, std::numeric_limits<float>::epsilon());
```